File: 4189-FindNthSmallestIntegerWithKOneBits/4189-FindNthSmallestIntegerWithKOneBits.py
# Last updated: 4/3/2026, 1:48:49 PM
class Solution:
    def __init__(self):
        self.C = [[0 for _ in range(61)] for _ in range(61)]
        self.precompute()
    
    def precompute(self):
        for i in range(61):
            self.C[i][0] = 1
            for j in range(1, i+1):
                self.C[i][j] = self.C[i-1][j-1] + self.C[i-1][j]
        
    def countKBits(self, x, k):
        if k < 0:
            return 0
        
        currK = k
        total = 0
        for i in range(60, -1, -1):
            if (x >> i) & 1:
                if currK <= i:
                    total += self.C[i][currK]
                currK -= 1
                if currK < 0:
                    break
        if currK == 0:
            total += 1
        return total

    def nthSmallest(self, n: int, k: int) -> int:
        if k < 0:
            return 0
        
        low = 1
        high = (1 << 60) - 1
        ans = high
        while low <= high:
            mid = low + (high - low) // 2
            if self.countKBits(mid, k) >= n:
                ans = mid
                high = mid-1
            else:
                low = mid+1
        
        return ans

File: 4189-FindNthSmallestIntegerWithKOneBits/4189-FindNthSmallestIntegerWithKOneBits.cpp
// Last updated: 4/3/2026, 1:48:53 PM
class Solution {
public:
    long long C[61][61];
    void precompute() {
        for (int i = 0; i <= 60; i++) {
            for (int j = 0; j <= 60; j++)
                C[i][j] = 0;
        }
        for (int i = 0; i <= 60; i++) {
            C[i][0] = 1;
            for (int j = 1; j <= i; j++)
                C[i][j] = C[i - 1][j - 1] + C[i - 1][j];
        }
    }
    long long countKBits(long long x, int k) {
        if (k < 0)
            return 0;

        int currK = k;
        long long total = 0;
        for (int i = 60; i >= 0; i--) {
            if ((x >> i) & 1) {
                if (currK <= i)
                    total += C[i][currK];
                currK--;
                if (currK < 0)
                    break;
            }
        }
        if (currK == 0)
            total++;
        return total;
    }
    long long nthSmallest(long long n, int k) {
        if (k < 0)
            return 0;

        precompute();
        long long low = 1;
        long long high = (1LL << 60) - 1;
        long long ans = high;
        while (low <= high) {
            long long mid = low + (high - low) / 2;
            if (countKBits(mid, k) >= n) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
};